jb/itch5/order_book_depth.hpp

```diff
@@ -122,29 +130,45 @@ class order_book {
    * @param side the side of the book to update
    * @param px the price of the order that was reduced
    * @param reduced_qty the quantity reduced in the order
-   * @returns true if the inside changed
-   */
+   * @return true if the inside changed
+   * Decrements order_book_ if a price (px) is erased from the map
+   *
+   * Check first if the px level exists. The use of find is ok since 
+   * the price level "has" to exist, and the iterator returned is used 
+   * to update the quantity (or erase it if qty is 0 or less)
+   * The following exceptions (no Throw executed) are logged as warnings
+   * into JOB_LOG
+   * EXC1: trying to reduce a non-existing price level
+   * EXC2: negative quantity in order book 
+   * EXC3: negative book_depth in order book
+  */
   template<typename book_side>
   bool handle_order_reduced(book_side& side, price4_t px, int reduced_qty) {
-    // Find the price level
-    // TODO() we create the price level if it does not exist, that is
-    // fairly inneficient ..
-    auto p = side.emplace(px, 0);
+    auto pf = side.find(px);
+    if (pf == side.end()) {
+      JB_LOG(warning) << "trying to reduce a non-existing price level";  // EXC1
+      return false;    // no change at the inside. Nothing else can be done
+    }
     // ... reduce the quantity ...
-    p.first->second -= reduced_qty;
-    if (p.first->second < 0) {
+    pf->second -= reduced_qty;
+    if (pf->second < 0) {
       // ... this is "Not Good[tm]", somehow we missed an order or
       // processed a delete twice ...
       // TODO() need a lot more details here...
-      JB_LOG(warning) << "negative quantity in order book";
+      JB_LOG(warning) << "negative quantity in order book";    // EXC2
     }
     // ... we may erase the element and invalidate the iterator, so
-    // compute if this was at the inside or not ...
-    bool r = p.first == side.begin();
-    if (p.first->second <= 0) {
-      side.erase(p.first);
+    // compute first if this was at the inside or not ...
+    bool change_inside = (pf == side.begin());
+    // now we can erase this element (pf.first) if qty <=0
+    if (pf->second <= 0) {
+      side.erase(pf);
+      if (book_depth_ == 0) 
```

Comment:
Please always use a block for if/else statements, missing braces are a source of trouble:

    if (foo)
        blah;
        blah2;

It looks like blah2 will happen only if foo is true, but it always runs.